In [42]:
import os
from pathlib import Path
from dotenv import find_dotenv

os.chdir(Path(find_dotenv()).parent)
print(Path.cwd().stem)

take_home_assignment


# Introduction

Basic EDA (exploratory data analysis).

# Summary

TODO: Brief summary of EDA

In [43]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from src.dataloader import XES3G5MDataModule, XES3G5MDataModuleConfig

---

# Load Data

In [44]:
data_module = XES3G5MDataModule(config=XES3G5MDataModuleConfig)
data_module.prepare_data()

In [47]:
stage = "fit"
data_module.setup(stage=stage)
data_module.seq_df_train

fold    uid                                          questions  \
0         0  11066  [3751, 3752, 3753, 3754, 1990, 3739, 3740, 374...   
1         0  11066  [95, 4294, 4294, 4300, 4300, 4301, 4302, 4308,...   
2         0  14482  [266, 268, 2203, 870, 2316, 2587, 2323, 2588, ...   
3         0  14482  [1280, 1046, 405, 404, 404, 4623, 4622, 4622, ...   
4         0  10532  [268, 266, 2203, 2588, 2587, 2323, 275, 811, 2...   
...     ...    ...                                                ...   
26707     3   3949  [2500, 2502, 2501, 2501, 2788, 3368, 3370, 336...   
26708     3  13641  [2208, 2333, 2334, 792, 791, 291, 293, 1946, 2...   
26709     3  13641  [1316, 1316, 2273, 2273, 1317, 1317, 1318, 131...   
26710     3  16217  [308, 1950, 1950, 2598, 819, 819, 181, 182, 22...   
26711     3  16217  [410, 414, 413, 408, 409, 407, 1283, 2265, 128...   

                                                concepts  \
0      [187, 187, 374, 187, 374, 188, 188, 228, 166, ...   
1      [63, 64, 102, 659, 376, 147, 147, 65, 64, 78, ...   
2      [9, 139, 140, 9, 496, 232, 231, 231, 146, 145,...   
3      [257, 32, 204, 204, 205, 206, 526, 206, 206, 2...   
4      [139, 9, 140, 231, 232, 231, 30, 30, 62, 65, 6...   
...                                                  ...   
26707  [335, 335, 187, 374, 81, 81, 336, 339, 336, 46...   
26708  [155, 155, 155, 303, 155, 55, 155, 155, 155, 1...   
26709  [390, 220, 220, 73, 24, 73, 66, 11, 17, 326, 7...   
26710  [160, 307, 55, 55, 307, 55, 108, 109, 109, 108...   
26711  [206, 208, 138, 207, 206, 206, 209, 138, 138, ...   

                                               responses  \
0      [1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, ...   
1      [0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, ...   
2      [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3      [0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, ...   
4      [1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, ...   
...                                                  ...   
26707  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
26708  [0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, ...   
26709  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
26710  [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, ...   
26711  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, ...   

                                              timestamps  \
0      [1595229836000, 1595233013000, 1595233687000, ...   
1      [1608890160000, 1608892236000, 1608892236000, ...   
2      [1595896054000, 1595896054000, 1595896054000, ...   
3      [1614471945000, 1614471945000, 1614471945000, ...   
4      [1595918717000, 1595918717000, 1595918717000, ...   
...                                                  ...   
26707  [1607668529000, 1607668529000, 1607668529000, ...   
26708  [1599300685000, 1599386011000, 1599386113000, ...   
26709  [1618728912000, 1618728912000, 1618728912000, ...   
26710  [1602075398000, 1602075705000, 1602075705000, ...   
26711  [1615028260000, 1616161054000, 1616161228000, ...   

                                             selectmasks  \
0      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
2      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
3      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
4      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
...                                                  ...   
26707  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
26708  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
26709  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
26710  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
26711  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   

                                               is_repeat  
0      [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, ...  
1      [0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
2      [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, ...  
3      [0,

❗ Long user sequences (> 200) are split into multiple rows, which are then treated as an independent data point. This means that the learning sequence is arbitrarily truncated (at 200), subsequent segments for the same user are treated as starting from scratch, hence introducing bias in the model.

Mitigation

* sliding window
* add positional encoding
* feature engineering / transfer information across segments
* increase max sequence length

In [23]:
data_module.seq_df_train.index

Int64Index([    0,     1,     2,     3,     4,     5,     6,     7,     8,
                9,
            ...
            26702, 26703, 26704, 26705, 26706, 26707, 26708, 26709, 26710,
            26711],
           dtype='int64', length=26712)

In [12]:
stage = "test"
data_module.setup(stage=stage)
test_dataloader = data_module.test_dataloader()

A user ID appears in multiple rows.

In [33]:
data_module.seq_df_train["uid"].value_counts()

8854     6
300      5
7460     4
9616     4
240      4
        ..
6882     2
15870    2
10636    2
4427     2
16217    2
Name: uid, Length: 11563, dtype: int64

In [34]:
data_module.seq_df_train.query("uid == 8854")

fold   uid                                          questions  \
11448     1  8854  [1859, 1864, 1865, 1863, 1860, 1862, 1861, 184...   
11449     1  8854  [1593, 1593, 1592, 1592, 1594, 1595, 1591, 159...   
11450     1  8854  [1608, 1610, 1615, 1607, 1688, 1686, 1689, 168...   
11451     1  8854  [1757, 1760, 1759, 1761, 94, 1925, 4563, 4560,...   
11452     1  8854  [1822, 1822, 1816, 1817, 1815, 1815, 1818, 181...   
11453     1  8854  [246, 248, 257, 1520, 2192, 2193, 2191, 250, 2...   

                                                concepts  \
11448  [214, 213, 155, 228, 155, 306, 155, 155, 5, 21...   
11449  [35, 73, 252, 73, 32, 33, 33, 73, 11, 165, 10,...   
11450  [420, 257, 257, 405, 431, 431, 83, 86, 86, 261...   
11451  [272, 65, 62, 64, 62, 62, 65, 65, 65, 679, 147...   
11452  [449, 455, 450, 167, 336, 100, 450, 61, 69, 22...   
11453  [133, 135, 99, 99, 132, 104, 133, 131, 137, 10...   

                                               responses  \
11448  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, ...   
11449  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
11450  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
11451  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
11452  [1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, ...   
11453  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   

                                              timestamps  \
11448  [1593350024000, 1593350024000, 1593350024000, ...   
11449  [1601088298000, 1601088298000, 1601088298000, ...   
11450  [1605589169000, 1605589169000, 1605589169000, ...   
11451  [1608906987000, 1608906987000, 1608906987000, ...   
11452  [1612097632000, 1612097632000, 1612097987000, ...   
11453  [1620393727000, 1620394354000, 1620394354000, ...   

                                             selectmasks  \
11448  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
11449  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
11450  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
11451  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
11452  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
11453  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   

                                               is_repeat  
11448  [0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, ...  
11449  [0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, ...  
11450  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
11451  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
11452  [0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, ...  
11453  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...

---

# Number of unique IDs

Used for the indexes of the nn.Embedding layers.

In [107]:
splits = ["train", "val", "test"]
columns = ["questions", "concepts"]
for col in columns:
    col_set = set()
    for split in splits:
        col_set |= set(getattr(data_module, f"seq_df_{split}")[col].explode().unique())
    print(f"Number of unique {col}: {len(col_set)}")

Number of unique questions: 7653
Number of unique concepts: 866


---

# Distributions Across Splits

TODO: Distributions of target across splits
TODO: Distributions of question IDs across splits
TODO: Distributions of `responses` per position and across splits


## Users

In [ ]:
assert len(set(data_module.seq_df_train.uid.unique()).intersection(data_module.seq_df_val.uid.unique())) == 0, "Train and validation sets should not overlap in user IDs"
assert len(set(data_module.seq_df_train.uid.unique()).intersection(data_module.seq_df_test.uid.unique())) == 0, "Train and test sets should not overlap in user IDs"
assert len(set(data_module.seq_df_test.uid.unique()).intersection(data_module.seq_df_val.uid.unique())) == 0, "Test and validation sets should not overlap in user IDs"

---

# Timestamps

## Monotonicity in a row

In [51]:
def is_non_decreasing(seq):
    return all(seq[i] <= seq[i+1] for i in range(len(seq) - 1))

def apply_selectmasks(row: pd.Series) -> list[int]:
    """Applies the `selectmasks` to the `timestamps`

    Args:
        row (pd.Series): The row of the dataframe containing the columns `timestamps` and `selectmasks`

    Returns:
        list[int]: A potentially truncated list of `timestamps`
    """
    return [ts for ts, sm in zip(row["timestamps"], row["selectmasks"]) if sm == 1]

In [ ]:
split = "train"
assert len(
    getattr(data_module, f"seq_df_{split}")
    .assign(is_non_decreasing_timestamps = lambda df: df.apply(apply_selectmasks, axis="columns").apply(is_non_decreasing))
    .query("~is_non_decreasing_timestamps")
) == 0, "There are some decreasing timestamps in a row"

In [56]:
split = "val"
assert len(
    getattr(data_module, f"seq_df_{split}")
    .assign(is_non_decreasing_timestamps = lambda df: df.apply(apply_selectmasks, axis="columns").apply(is_non_decreasing))
    .query("~is_non_decreasing_timestamps")
) == 0, "There are some decreasing timestamps in a row"

In [57]:
split = "test"
assert len(
    getattr(data_module, f"seq_df_{split}")
    .assign(is_non_decreasing_timestamps = lambda df: df.apply(apply_selectmasks, axis="columns").apply(is_non_decreasing))
    .query("~is_non_decreasing_timestamps")
) == 0, "There are some decreasing timestamps in a row"

## Monotonicity across rows

In [69]:
def non_decreasing_within_a_group(df: pd.DataFrame) -> bool:
    """Users with a sequence larger than 200 are split into multiple rows.
    Check that the last value in the `timestamps` list is less than or equal to the 
    the first value in the `timestamps` list in the next row.

    Args:
        df (pd.DataFrame): The dataframe containing the columns `timestamps` and `selectmasks`

    Returns:
        bool: True if all rows agree to the logic.
    """
    timestamps = df["timestamps"].tolist()
    comparisons = []
    for i, seq in enumerate(timestamps[:-1]):
        comparisons.append(seq[-1] <= timestamps[i+1][0])
    return all(comparisons)

In [71]:
split = "train"
assert all(
    getattr(data_module, f"seq_df_{split}")
    .groupby("uid")
    .apply(non_decreasing_within_a_group)
), f"Some users in the {split} split are not sorted."

In [72]:
split = "val"
assert all(
    getattr(data_module, f"seq_df_{split}")
    .groupby("uid")
    .apply(non_decreasing_within_a_group)
), f"Some users in the {split} split are not sorted."

In [73]:
split = "test"
assert all(
    getattr(data_module, f"seq_df_{split}")
    .groupby("uid")
    .apply(non_decreasing_within_a_group)
), f"Some users in the {split} split are not sorted."

---

# Embeddings

## Embedding Dimensions

In [58]:
print(data_module.question_embeddings.shape)
print(data_module.concept_embeddings.shape)

(7652, 768)
(1175, 768)


## Match Indexes to Questions

Potentially exclude questions and concepts that are missing the embeddings.

In [31]:
train_dataloader = data_module.train_dataloader()

In [32]:
train_dataloader_iter = iter(train_dataloader)
train_batch = next(train_dataloader_iter)

In [33]:
train_batch["questions"].size(1)

200

In [37]:
train_batch["questions"].size(0)

64

In [16]:
train_batch["questions"].shape

torch.Size([64, 200])

In [17]:
train_batch["question_embeddings"].shape

torch.Size([64, 200, 768])

In [19]:
train_batch["question_embeddings"][0].shape

torch.Size([200, 768])

In [11]:
train_batch.keys()

dict_keys(['questions', 'concepts', 'question_embeddings', 'concept_embeddings', 'selectmasks', 'responses'])

In [82]:
train_dataloader.dataset[0]

{'questions': tensor([3751, 3752, 3753, 3754, 1990, 3739, 3740, 3742, 3756, 3756, 3757, 3600,
         3758, 3760, 3761, 3925, 3928, 1916, 1914, 3764, 3763, 3767, 3766, 3768,
         3931, 3933, 3772, 5749, 4664, 5750, 5751, 5752, 5754, 5755,    6, 3934,
         3935, 3936, 3937, 3941, 3943, 3946,    7, 3162, 3163, 3264, 3951, 3952,
         3954, 3955, 3956, 3958, 3959, 2053, 3961, 3960, 3963, 3968, 3966, 3966,
         3967, 1875, 1875,  513, 2047, 2048, 2048, 3971, 3972, 3973, 3974, 3977,
         3978,  524, 3979,   27,   27, 3127, 3127, 3982, 3980, 3981, 3983, 3984,
         3987, 3989, 3995, 3993, 3994, 3996, 3997, 3998, 3999, 4001, 4000, 4002,
         4003, 4004, 4005, 4006, 4007, 4008, 4009, 4010, 4013,   35, 5525, 5527,
         5526, 5836, 5529, 5528, 5837, 5530, 5531, 5838, 5532, 5533, 5534, 5535,
         5839, 5841, 5538, 2064, 2063, 4014, 4016, 4015, 4291, 4017, 4019, 4021,
         4022, 2071, 4023, 1440,  561, 4025, 4026, 2074, 4027, 4028, 4029, 3171,
          554, 

In [85]:
train_dataloader.dataset[1]

{'questions': tensor([  95, 4294, 4294, 4300, 4300, 4301, 4302, 4308,  115,  130,  136, 1927,
         2125, 2127, 2128, 2130, 2129, 2131, 2136, 2134, 2135, 2139,  670,  671,
          163,  163, 1928, 1929, 3230,  678,  679,  680,  684,  687,  689,  689,
          688,  688,  167, 4091, 4091, 4092, 4092, 2143, 2143, 2146, 2146, 4093,
         4094, 2147, 2147, 2148, 2148, 2148, 4941,  718,  188,  192,  193,  191,
         5540, 5842, 5541, 5191,  230,  232,  231, 4120, 4120, 4322, 4322, 4121,
         4122, 4123, 4323, 4124,  756, 4125,  242,  244,  243,  767,  768, 2189,
          770,  770,  771,  771, 3642, 3642,  772, 5849, 5849, 5542, 4133,  260,
         1935, 4135, 4136, 1936, 1533, 4140, 4143, 4144, 4145,   -1,   -1,   -1,
           -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,
           -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,
           -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,
           -1, 

In [84]:
data_module.seq_df_train.query("uid == 11066")

fold    uid                                          questions  \
0     0  11066  [3751, 3752, 3753, 3754, 1990, 3739, 3740, 374...   
1     0  11066  [95, 4294, 4294, 4300, 4300, 4301, 4302, 4308,...   

                                            concepts  \
0  [187, 187, 374, 187, 374, 188, 188, 228, 166, ...   
1  [63, 64, 102, 659, 376, 147, 147, 65, 64, 78, ...   

                                           responses  \
0  [1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, ...   
1  [0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, ...   

                                          timestamps  \
0  [1595229836000, 1595233013000, 1595233687000, ...   
1  [1608890160000, 1608892236000, 1608892236000, ...   

                                         selectmasks  \
0  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   

                                           is_repeat  
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, ...  
1  [0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...

---

# Correctness Distribution On the First Segment vs Subsequent Segments

To test if arbitrarily splitting the sequence into chunks of 200 could introduce bias, check if the proportions of correct answers for questions in the first segment differs from those in the subsequent segments. For example, if question 4294 answered in segment 1 (within sequence 1-200) has a probability of being correct of 50%, and in segment 2 (within sequence 201-400) has a probability of being correct of 80%, this likely means that the student has learned to generalise the knowledge gained in the first segment, hence they were able to answer 4294 with a high probability of success. If this is the case, then treating subsequent segments as if they are independent (i.e. not having any history) would introduct bias to the model.

---

# OOV (Out Of Vocabulary) IDs

## Question IDs

In [14]:
len(set(data_module.seq_df_val.questions.explode().unique()) - set(data_module.seq_df_train.questions.explode().unique()))

101

In [15]:
len(set(data_module.seq_df_test.questions.explode().unique()) - set(data_module.seq_df_train.questions.explode().unique()))

47

## Knowledge Component IDs

---

# Appendix - Question-Level vs KC-Level

In [ ]:
# Question level data
n_rows = 1000
q_level_df = pd.read_csv("data/XES3G5M/question_level/train_valid_sequences_quelevel.csv", nrows=n_rows)
q_level_df


fold    uid                                          questions  \
0       0  11066  3751,3752,3753,3754,1990,3739,3740,3742,3756,3...   
1       0  11066  1927,2125,2127,2128,2130,2129,2131,2136,2134,2...   
2       0  14482  266,268,2203,870,2316,2587,2323,2588,5194,868,...   
3       0  14482  4929,1073,424,1074,439,1079,1078,440,1318,4931...   
4       0  10532  268,266,2203,2588,2587,2323,275,811,2205,277,2...   
..    ...    ...                                                ...   
995     0  17005  2312,12,1129,1131,1130,267,2314,266,2309,863,8...   
996     0  17005  1317,1316,442,2392,429,1336,1100,1102,1101,109...   
997     0   6010  2323,2322,274,273,275,812,811,278,277,276,290,...   
998     0   6010  1276,1270,1053,2365,2366,403,404,405,405,404,4...   
999     0  11640  3751,3752,3753,3916,3918,1990,3740,3739,3741,3...   

                                              concepts  \
0    187,187,374,187,374,188,188,228,166_170,221,40...   
1    468,78,78,78,78,78,78,378,378,378,82,78,78,64_...   
2    9,139,140,9,496,232,231,231,146,145_231,31,65,...   
3    214,214,214,214,24,23,23,220,66_11_17_326,359,...   
4    139,9,140,231,232,231,30,30,62_65,62,65,152,15...   
..                                                 ...   
995  139_9,9,365,66_365,366,141,139,9,365,142,223,1...   
996  24_73,390_220,140,502,213_217,238_180,180,176,...   
997  231,232,119,146,30,30,30,147,62,65,54,470,51,1...   
998  257,257,252,257,257,204,204_205,204,204,204_20...   
999  187,187,374,188,188,374,188,188,142,228,228,21...   

                                             responses  \
0    1,1,1,0,1,1,1,1,0,0,1,1,1,1,1,1,1,1,1,0,1,1,1,...   
1    1,1,1,1,1,1,1,1,0,0,0,1,1,1,1,0,1,1,1,1,1,1,1,...   
2    0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,1,1,1,...   
3    1,1,1,1,1,1,1,1,1,0,0,1,0,1,1,1,1,1,0,0,0,0,0,...   
4    1,0,0,0,1,0,1,1,1,1,0,1,0,1,0,0,0,1,1,1,0,1,0,...   
..                                                 ...   
995  1,1,1,1,1,1,1,1,1,1,0,1,0,0,0,0,1,1,1,1,1,1,1,...   
996  0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,...   
997  1,0,1,1,1,1,1,1,1,0,0,0,0,1,1,0,0,1,1,1,1,1,1,...   
998  1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,1,1,1,1,1,1,...   
999  1,0,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,0,1,1,1,1,1,...   

                                            timestamps  \
0    1595229836000,1595233013000,1595233687000,1595...   
1    1611827809000,1611829702000,1611832165000,1611...   
2    1595896054000,1595896054000,1595896054000,1595...   
3    1616289797000,1616300013000,1616300013000,1616...   
4    1595918717000,1595918717000,1595918717000,1596...   
..                                                 ...   
995  1594526920000,1594526920000,1594563373000,1594...   
996  1618036139000,1618036139000,1618641549000,1618...   
997  1596414359000,1596414359000,1596500722000,1596...   
998  1612868724000,1612868724000,1612868816000,1612...   
999  1597639606000,1597642795000,1597643297000,1597...   

                                           selectmasks  
0    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  
1    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  
2    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  
3    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  
4    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  
..                                                 ...  
995  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  
996  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  
997  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  
998  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  
999  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  

[1000 rows x 7 columns]

In [56]:
q_long_df = pd.DataFrame(
    q_level_df
    .query("uid == 11066")
    .drop(columns=["fold", "uid"])
    .iloc[0]
    .apply(lambda x: x.split(","))
    .tolist(),
    index=["questions", "concepts", "responses", "timestamps", "selectmasks"]
).T

q_long_df

questions concepts responses     timestamps selectmasks
0        3751      187         1  1595229836000           1
1        3752      187         1  1595233013000           1
2        3753      374         1  1595233687000           1
3        3754      187         0  1595236010000           1
4        1990      374         1  1595314541000           1
..        ...      ...       ...            ...         ...
195      4302      147         0  1608896293000           1
196      4308       65         0  1608899418000           1
197       115       64         1  1609315433000           1
198       130       78         1  1611826578000           1
199       136       81         0  1611827809000           1

[200 rows x 5 columns]

In [21]:
# KC level data
kc_level_df = pd.read_csv("data/XES3G5M/kc_level/train_valid_sequences.csv", nrows=n_rows)
kc_level_df

fold    uid                                          questions  \
0       0  11066  3751,3752,3753,3754,1990,3739,3740,3742,3756,3...   
1       0  11066  95,4294,4294,4300,4300,4301,4302,4308,115,130,...   
2       0  14482  266,268,2203,870,2316,2587,2323,2588,5194,868,...   
3       0  14482  1280,1046,405,404,404,4623,4622,4622,4624,4954...   
4       0  10532  268,266,2203,2588,2587,2323,275,811,2205,2205,...   
..    ...    ...                                                ...   
995     0   5709  1129,1131,1131,1130,263,264,265,267,268,266,11...   
996     0   5709  403,404,404,405,2367,2380,2381,413,412,415,416...   
997     0   6917  1991,1992,2003,2001,2004,2002,467,467,467,2012...   
998     0   6917  650,651,2120,653,653,2122,2122,654,655,656,657...   
999     0    207  2312,2312,12,2328,2328,2204,274,2591,826,2318,...   

                                              concepts  \
0    187,187,374,187,374,188,188,228,166,170,221,40...   
1    63,64,102,659,376,147,147,65,64,78,81,468,78,7...   
2    9,139,140,9,496,232,231,231,146,145,231,31,65,...   
3    257,32,204,204,205,206,526,206,206,206,206,207...   
4    139,9,140,231,232,231,30,30,62,65,62,65,152,15...   
..                                                 ...   
995  365,66,365,366,139,9,140,141,139,9,142,494,301...   
996  204,204,205,204,501,138,300,138,138,209,138,20...   
997  213,214,229,229,229,229,233,234,235,233,234,23...   
998  102,147,147,272,147,272,147,65,65,65,65,86,447...   
999  139,9,9,497,488,119,119,30,308,231,232,148,46,...   

                                             responses  \
0    1,1,1,0,1,1,1,1,0,0,0,1,1,1,1,1,1,1,1,1,0,1,1,...   
1    0,1,1,1,1,1,0,0,1,1,0,1,1,1,1,1,1,1,1,0,0,0,1,...   
2    0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,1,...   
3    0,1,1,1,1,1,1,1,1,0,0,0,1,1,1,1,1,0,0,1,1,1,1,...   
4    1,0,0,0,1,0,1,1,1,1,1,0,1,0,1,0,0,0,1,1,1,0,1,...   
..                                                 ...   
995  1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,1,0,1,...   
996  1,1,1,1,1,1,0,1,1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,...   
997  1,0,0,1,0,0,1,1,1,0,0,0,1,1,1,1,1,0,1,0,1,1,1,...   
998  1,0,1,0,0,0,0,0,1,0,0,1,1,1,1,1,1,1,1,1,1,0,0,...   
999  1,1,1,0,0,1,1,1,1,1,1,0,0,1,1,1,0,1,1,1,1,1,1,...   

                                            timestamps  \
0    1595229836000,1595233013000,1595233687000,1595...   
1    1608890160000,1608892236000,1608892236000,1608...   
2    1595896054000,1595896054000,1595896054000,1595...   
3    1614471945000,1614471945000,1614471945000,1614...   
4    1595918717000,1595918717000,1595918717000,1596...   
..                                                 ...   
995  1594428907000,1594428907000,1594428907000,1594...   
996  1614992125000,1614992125000,1614992125000,1614...   
997  1596508399000,1596508777000,1596675389000,1596...   
998  1608947653000,1608948858000,1608949075000,1608...   
999  1593434484000,1593434484000,1593434484000,1593...   

                                           selectmasks  \
0    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
1    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
2    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
3    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
4    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
..                                                 ...   
995  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
996  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
997  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
998  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
999  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   

                                             is_repeat  
0    0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,...  
1    0,0,1,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...  
2    0,0,0,0,0,0,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,...  
3    0,0,0,0,1,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...  
4    0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,...  
..                    

In [50]:
long_df = pd.DataFrame(
    kc_level_df
    .query("uid == 11066")
    .drop(columns=["fold", "uid"])
    .iloc[0]
    .apply(lambda x: x.split(","))
    .tolist(),
    index=["questions", "concepts", "responses", "timestamps", "selectmasks", "is_repeat"]
).T

long_df

questions concepts responses     timestamps selectmasks is_repeat
0        3751      187         1  1595229836000           1         0
1        3752      187         1  1595233013000           1         0
2        3753      374         1  1595233687000           1         0
3        3754      187         0  1595236010000           1         0
4        1990      374         1  1595314541000           1         0
..        ...      ...       ...            ...         ...       ...
195      4076       57         0  1607689417000           1         0
196      3597       55         1  1607765000000           1         0
197      3092       46         1  1607765147000           1         0
198        97       56         1  1608890160000           1         0
199        96       58         0  1608890160000           1         0

[200 rows x 6 columns]

In [54]:
(
    long_df
    .assign(dup_questions = lambda df: df["questions"].duplicated(keep=False))
)

questions concepts responses     timestamps selectmasks is_repeat  \
0        3751      187         1  1595229836000           1         0   
1        3752      187         1  1595233013000           1         0   
2        3753      374         1  1595233687000           1         0   
3        3754      187         0  1595236010000           1         0   
4        1990      374         1  1595314541000           1         0   
..        ...      ...       ...            ...         ...       ...   
195      4076       57         0  1607689417000           1         0   
196      3597       55         1  1607765000000           1         0   
197      3092       46         1  1607765147000           1         0   
198        97       56         1  1608890160000           1         0   
199        96       58         0  1608890160000           1         0   

     dup_questions  
0            False  
1            False  
2            False  
3            False  
4            False  
..             ...  
195          False  
196          False  
197          False  
198          False  
199          False  

[200 rows x 7 columns]

In [30]:
data_module.seq_df_train.iloc[1]["selectmasks"]

array([ 1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1])

In [38]:
data_module.seq_df_train.iloc[1]["questions"]

array([  95, 4294, 4294, 4300, 4300, 4301, 4302, 4308,  115,  130,  136,
       1927, 2125, 2127, 2128, 2130, 2129, 2131, 2136, 2134, 2135, 2139,
        670,  671,  163,  163, 1928, 1929, 3230,  678,  679,  680,  684,
        687,  689,  689,  688,  688,  167, 4091, 4091, 4092, 4092, 2143,
       2143, 2146, 2146, 4093, 4094, 2147, 2147, 2148, 2148, 2148, 4941,
        718,  188,  192,  193,  191, 5540, 5842, 5541, 5191,  230,  232,
        231, 4120, 4120, 4322, 4322, 4121, 4122, 4123, 4323, 4124,  756,
       4125,  242,  244,  243,  767,  768, 2189,  770,  770,  771,  771,
       3642, 3642,  772, 5849, 5849, 5542, 4133,  260, 1935, 4135, 4136,
       1936, 1533, 4140, 4143, 4144, 4145,   -1,   -1,   -1,   -1,   -1,
         -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,
         -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,
         -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,   -1,
         -1,   -1,   -1,   -1,   -1,   -1,   -1,   

---

# Check if question_ids are zero-based

Yes, there are 0-valued question indexes. Note the -1 is for padding. This means that the padding can't be changed to 0.

In [41]:
sorted(data_module.seq_df_train.questions.explode().unique().tolist())

[-1,
 0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 

# Delete

In [51]:
import torch

In [48]:
train_dataloader = data_module.train_dataloader()
train_dataloader_iter = iter(train_dataloader)
train_batch = next(train_dataloader_iter)

In [56]:
question_ids = train_batch["questions"]
question_ids

tensor([[1130, 1129, 1131,  ...,  391, 1223, 1228],
        [3452,  352,  349,  ...,   -1,   -1,   -1],
        [2001, 2002, 2004,  ..., 2084,  595, 2085],
        ...,
        [ 609,  610,  610,  ...,   -1,   -1,   -1],
        [ 562,  563,  574,  ..., 2144, 2145, 2145],
        [1387,  883,  312,  ..., 1224, 1225, 1225]])

In [55]:
question_ids.device

device(type='cpu')

In [54]:
# question_ids: shape (batch_size, seq_len)
# self.question_embedding_idx: the index you want to use for padding

question_ids = torch.where(
    question_ids == -1,
    torch.tensor(9999, device=question_ids.device, dtype=question_ids.dtype),
    question_ids
)
question_ids

tensor([[1130, 1129, 1131,  ...,  391, 1223, 1228],
        [3452,  352,  349,  ..., 9999, 9999, 9999],
        [2001, 2002, 2004,  ..., 2084,  595, 2085],
        ...,
        [ 609,  610,  610,  ..., 9999, 9999, 9999],
        [ 562,  563,  574,  ..., 2144, 2145, 2145],
        [1387,  883,  312,  ..., 1224, 1225, 1225]])

In [57]:
question_ids = torch.where(
            question_ids == -1,
            torch.full_like(question_ids, 9999),
            question_ids,
        )
question_ids

tensor([[1130, 1129, 1131,  ...,  391, 1223, 1228],
        [3452,  352,  349,  ..., 9999, 9999, 9999],
        [2001, 2002, 2004,  ..., 2084,  595, 2085],
        ...,
        [ 609,  610,  610,  ..., 9999, 9999, 9999],
        [ 562,  563,  574,  ..., 2144, 2145, 2145],
        [1387,  883,  312,  ..., 1224, 1225, 1225]])

In [61]:
responses = train_batch["responses"]
responses

tensor([[1, 1, 1,  ..., 1, 1, 1],
        [0, 0, 1,  ..., 0, 0, 0],
        [1, 0, 0,  ..., 0, 1, 1],
        ...,
        [1, 0, 0,  ..., 0, 0, 0],
        [1, 1, 1,  ..., 1, 1, 1],
        [1, 1, 1,  ..., 0, 1, 1]])

In [64]:
num_questions = 3000

In [65]:
num_questions * responses

tensor([[3000, 3000, 3000,  ..., 3000, 3000, 3000],
        [   0,    0, 3000,  ...,    0,    0,    0],
        [3000,    0,    0,  ...,    0, 3000, 3000],
        ...,
        [3000,    0,    0,  ...,    0,    0,    0],
        [3000, 3000, 3000,  ..., 3000, 3000, 3000],
        [3000, 3000, 3000,  ...,    0, 3000, 3000]])

In [67]:
interaction_ids

tensor([[1330, 1329, 1331,  ...,  591, 1423, 1428],
        [3452,  352,  549,  ..., 9999, 9999, 9999],
        [2201, 2002, 2004,  ..., 2084,  795, 2285],
        ...,
        [ 809,  610,  610,  ..., 9999, 9999, 9999],
        [ 762,  763,  774,  ..., 2344, 2345, 2345],
        [1587, 1083,  512,  ..., 1224, 1425, 1425]])

In [68]:
num_questions = 3000

interaction_ids = question_ids + num_questions * responses
interaction_ids

tensor([[4130, 4129, 4131,  ..., 3391, 4223, 4228],
        [3452,  352, 3349,  ..., 9999, 9999, 9999],
        [5001, 2002, 2004,  ..., 2084, 3595, 5085],
        ...,
        [3609,  610,  610,  ..., 9999, 9999, 9999],
        [3562, 3563, 3574,  ..., 5144, 5145, 5145],
        [4387, 3883, 3312,  ..., 1224, 4225, 4225]])

In [74]:
seq_len = 200
batch_size = 64
position_ids = torch.arange(seq_len).unsqueeze(0).repeat(batch_size, 1)
print(position_ids)
print(position_ids.shape)

tensor([[  0,   1,   2,  ..., 197, 198, 199],
        [  0,   1,   2,  ..., 197, 198, 199],
        [  0,   1,   2,  ..., 197, 198, 199],
        ...,
        [  0,   1,   2,  ..., 197, 198, 199],
        [  0,   1,   2,  ..., 197, 198, 199],
        [  0,   1,   2,  ..., 197, 198, 199]])
torch.Size([64, 200])


In [76]:
selectmasks

tensor([[ 1.,  1.,  1.,  ...,  1.,  1.,  1.],
        [ 1.,  1.,  1.,  ..., -1., -1., -1.],
        [ 1.,  1.,  1.,  ...,  1.,  1.,  1.],
        ...,
        [ 1.,  1.,  1.,  ..., -1., -1., -1.],
        [ 1.,  1.,  1.,  ...,  1.,  1.,  1.],
        [ 1.,  1.,  1.,  ...,  1.,  1.,  1.]])

In [77]:
selectmasks = train_batch["selectmasks"]
position_padding_idx = 200
position_ids = torch.where(
            selectmasks == -1,
            torch.full_like(position_ids, position_padding_idx),
            position_ids,
        )
position_ids

tensor([[  0,   1,   2,  ..., 197, 198, 199],
        [  0,   1,   2,  ..., 200, 200, 200],
        [  0,   1,   2,  ..., 197, 198, 199],
        ...,
        [  0,   1,   2,  ..., 200, 200, 200],
        [  0,   1,   2,  ..., 197, 198, 199],
        [  0,   1,   2,  ..., 197, 198, 199]])

In [83]:
position_ids.shape

torch.Size([64, 200])

In [109]:
len(train_dataloader.dataset)//64

417

In [117]:
train_dataloader.dataset.__len__()

26712

In [121]:
train_df = data_module.datasets["sequence"]["train"].to_pandas()
train_df

fold    uid                                          questions  \
0         0  11066  [3751, 3752, 3753, 3754, 1990, 3739, 3740, 374...   
1         0  11066  [95, 4294, 4294, 4300, 4300, 4301, 4302, 4308,...   
2         0  14482  [266, 268, 2203, 870, 2316, 2587, 2323, 2588, ...   
3         0  14482  [1280, 1046, 405, 404, 404, 4623, 4622, 4622, ...   
4         0  10532  [268, 266, 2203, 2588, 2587, 2323, 275, 811, 2...   
...     ...    ...                                                ...   
33392     4  13281  [1160, 1161, 2232, 2342, 2343, 2229, 1958, 351...   
33393     4  13281  [1303, 1306, 1305, 1305, 1300, 1298, 1299, 239...   
33394     4   9822  [1130, 1129, 1131, 1131, 264, 263, 265, 267, 2...   
33395     4   9822  [1193, 1192, 981, 982, 365, 1964, 1965, 1966, ...   
33396     4   9822  [407, 409, 1281, 1283, 415, 416, 418, 418, 417...   

                                                concepts  \
0      [187, 187, 374, 187, 374, 188, 188, 228, 166, ...   
1      [63, 64, 102, 659, 376, 147, 147, 65, 64, 78, ...   
2      [9, 139, 140, 9, 496, 232, 231, 231, 146, 145,...   
3      [257, 32, 204, 204, 205, 206, 526, 206, 206, 2...   
4      [139, 9, 140, 231, 232, 231, 30, 30, 62, 65, 6...   
...                                                  ...   
33392  [166, 166, 168, 359, 265, 203, 168, 184, 176, ...   
33393  [24, 389, 388, 205, 23, 24, 23, 223, 222, 222,...   
33394  [366, 365, 66, 365, 9, 139, 140, 141, 9, 139, ...   
33395  [335, 335, 334, 335, 81, 335, 471, 188, 471, 4...   
33396  [206, 206, 138, 209, 209, 138, 210, 211, 208, ...   

                                               responses  \
0      [1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, ...   
1      [0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, ...   
2      [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3      [0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, ...   
4      [1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, ...   
...                                                  ...   
33392  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
33393  [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, ...   
33394  [1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, ...   
33395  [1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
33396  [1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, ...   

                                              timestamps  \
0      [1595229836000, 1595233013000, 1595233687000, ...   
1      [1608890160000, 1608892236000, 1608892236000, ...   
2      [1595896054000, 1595896054000, 1595896054000, ...   
3      [1614471945000, 1614471945000, 1614471945000, ...   
4      [1595918717000, 1595918717000, 1595918717000, ...   
...                                                  ...   
33392  [1604586260000, 1604586260000, 1604586260000, ...   
33393  [1618059571000, 1618059571000, 1618059571000, ...   
33394  [1594460706000, 1594460706000, 1594460706000, ...   
33395  [1607699449000, 1607699449000, 1607742804000, ...   
33396  [1615696235000, 1615696235000, 1616215619000, ...   

                                             selectmasks  \
0      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
2      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
3      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
4      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
...                                                  ...   
33392  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
33393  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
33394  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
33395  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
33396  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   

                                               is_repeat  
0      [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, ...  
1      [0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
2      [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, ...  
3      [0,

In [ ]:
data_module.

0    6669
1    6669
Name: fold, dtype: int64

In [127]:
train_df.loc[idx]

fold    uid                                          questions  \
0         0  11066  [3751, 3752, 3753, 3754, 1990, 3739, 3740, 374...   
1         0  11066  [95, 4294, 4294, 4300, 4300, 4301, 4302, 4308,...   
2         0  14482  [266, 268, 2203, 870, 2316, 2587, 2323, 2588, ...   
3         0  14482  [1280, 1046, 405, 404, 404, 4623, 4622, 4622, ...   
4         0  10532  [268, 266, 2203, 2588, 2587, 2323, 275, 811, 2...   
...     ...    ...                                                ...   
13333     1    584  [270, 271, 269, 811, 812, 276, 277, 278, 1942,...   
13334     1    584  [1174, 1174, 1180, 1177, 1178, 1172, 1173, 120...   
13335     1    584  [431, 430, 432, 433, 435, 435, 437, 436, 1318,...   
13336     1   2050  [266, 268, 269, 270, 271, 2323, 2322, 1937, 81...   
13337     1   2050  [1273, 1280, 2365, 2366, 1051, 1053, 1052, 366...   

                                                concepts  \
0      [187, 187, 374, 187, 374, 188, 188, 228, 166, ...   
1      [63, 64, 102, 659, 376, 147, 147, 65, 64, 78, ...   
2      [9, 139, 140, 9, 496, 232, 231, 231, 146, 145,...   
3      [257, 32, 204, 204, 205, 206, 526, 206, 206, 2...   
4      [139, 9, 140, 231, 232, 231, 30, 30, 62, 65, 6...   
...                                                  ...   
13333  [143, 144, 142, 30, 30, 65, 62, 147, 470, 51, ...   
13334  [329, 57, 57, 179, 170, 177, 179, 183, 76, 187...   
13335  [214, 214, 214, 213, 218, 213, 219, 214, 66, 1...   
13336  [9, 139, 142, 143, 144, 231, 232, 232, 30, 30,...   
13337  [260, 257, 257, 257, 33, 252, 32, 420, 257, 25...   

                                               responses  \
0      [1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, ...   
1      [0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, ...   
2      [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3      [0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, ...   
4      [1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, ...   
...                                                  ...   
13333  [1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, ...   
13334  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
13335  [1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, ...   
13336  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
13337  [1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, ...   

                                              timestamps  \
0      [1595229836000, 1595233013000, 1595233687000, ...   
1      [1608890160000, 1608892236000, 1608892236000, ...   
2      [1595896054000, 1595896054000, 1595896054000, ...   
3      [1614471945000, 1614471945000, 1614471945000, ...   
4      [1595918717000, 1595918717000, 1595918717000, ...   
...                                                  ...   
13333  [1595670319000, 1595670319000, 1595670319000, ...   
13334  [1607520549000, 1607520549000, 1607520549000, ...   
13335  [1616933033000, 1616933033000, 1616933033000, ...   
13336  [1596362103000, 1596362103000, 1596448506000, ...   
13337  [1611239377000, 1611240208000, 1611240208000, ...   

                                             selectmasks  \
0      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
2      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
3      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
4      [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
...                                                  ...   
13333  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
13334  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
13335  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
13336  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
13337  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   

                                               is_repeat  
0      [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, ...  
1      [0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
2      [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, ...  
3      [0,

In [131]:
data_module

371